### PR TITLE
Fs async

### PR DESCRIPTION
**Pull request by joshuef**

### CODE REVIEW

**Review by Yoga07 — DISMISSED** [review]

**Review by Yoga07 — APPROVED** [review]

File: src/client/client_api/map_apis.rs [hunk 1]
```diff
@@ -887,16 +887,24 @@ mod tests {
...
         let test_data_name = XorName(rand::random());
-        let address = client
-            .store_seq_map(test_data_name, 15000u64, random_pk, None, Some(permissions))
-            .await?;
+        let address = retry_loop!(client.store_seq_map(
+            test_data_name,
+            15000u64,
+            random_pk,
+            None,
+            Some(permissions.clone())
```
Comment (grumbach) [inline src/client/client_api/map_apis.rs:895]: Just thinking out loud, `clone()` in a loop, wouldn't it be possible to pass a ref?

File: src/client/client_api/sequence_apis.rs [hunk 2]
```diff
@@ -783,27 +783,18 @@ mod tests {
         let client = create_test_client().await?;
         let owner = client.public_key();
         let perms = BTreeMap::<PublicKey, SequencePrivatePermissions>::new();
-        let sequence_address = client
-            .store_private_sequence(None, name, tag, owner, perms)
-            .await?;
-
-        let mut balance_before_delete = client.get_balance().await?;
+        let sequence_address =
+            retry_loop!(client.store_private_sequence(None, name, tag, owner, perms.clone()));
```
Comment (grumbach) [inline src/client/client_api/sequence_apis.rs:787]: Same comment about `clone()`

File: src/client/client_api/sequence_apis.rs [hunk 3]
```diff
@@ -783,27 +783,18 @@ mod tests {
...
 
-        while balance_before_delete == Token::from_str("0")?
-            || balance_before_delete == Token::from_str("10")?
-        {
-            sleep(Duration::from_millis(200)).await;
-            balance_before_delete = client.get_balance().await?;
-        }
+        let _ =
+            retry_loop_for_pattern!(client.get_balance(), Ok(bal) if *bal != Token::from_str("0")?);
+        let balance_before_delete = retry_loop_for_pattern!(client.get_balance(), Ok(bal) if *bal != Token::from_str("10")?)?;
```
Comment (grumbach) [inline src/client/client_api/sequence_apis.rs:791]: This is one more `get_balance()` than the impl before, I suppose `get_balance()` problably has no side effects, but could it change between the 2 calls?

File: src/client/client_api/sequence_apis.rs [hunk 4]
```diff
@@ -875,38 +852,22 @@ mod tests {
...
+            retry_loop!(client.get_sequence_private_permissions_for_user(address, owner));
 
         assert!(user_perms.is_allowed(SequenceAction::Read));
         assert!(user_perms.is_allowed(SequenceAction::Append));
 
         let mut seq_res = client
             .get_sequence_permissions(address, SequenceUser::Key(owner))
             .await;
+
         while seq_res.is_err() {
```
Comment (grumbach) [inline src/client/client_api/sequence_apis.rs:869]: Why not use the `retry_loop!` marco this time?

File: src/node/data_store/tests.rs [hunk 5]
```diff
@@ -161,10 +161,31 @@ async fn delete() -> Result<()> {
             value: data.clone(),
         };
         data_store.put(the_data).await?;
+
+        let mut used_space = data_store.total_used_space().await;
+
+        while used_space != *size {
+            used_space = data_store.total_used_space().await;
+            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
+        }
```
Comment (grumbach) [inline src/node/data_store/tests.rs]:
```suggestion
        loop {
            let used_space = data_store.total_used_space().await;
            if used_space == *size {
                break;
            }
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
        }
```
No mut, only one call. What do you think?

File: src/node/data_store/tests.rs [hunk 6]
```diff
@@ -161,10 +161,31 @@ async fn delete() -> Result<()> {
...
+        assert!(!data_store.has(&the_data.id).await);
+
+        let mut used_space = data_store.total_used_space().await;
+
+        while used_space != 0 {
+            used_space = data_store.total_used_space().await;
+            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
+        }
+
```
Comment (grumbach) [inline src/node/data_store/tests.rs]:
```suggestion
        loop {
            let used_space = data_store.total_used_space().await;
            if used_space == 0 {
                break;
            }
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
        }
```

File: src/node/data_store/tests.rs [hunk 7]
```diff
@@ -211,8 +232,16 @@ async fn overwrite_value() -> Result<()> {
                 value: data.clone(),
             })
             .await?;
+
+        let mut used_space = data_store.total_used_space().await;
+
+        while used_space != size {
+            used_space = data_store.total_used_space().await;
+            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
+        }
```
Comment (grumbach) [inline src/node/data_store/tests.rs:249]:
```suggestion
        loop {
            let used_space = data_store.total_used_space().await;
            if used_space == size {
                break;
            }
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
        }
```

**Review by grumbach — COMMENTED** [review]
A couple of comments and code suggestions but nothing game changing. Maddening amount of await additions! 🥇

File: src/client/client_api/map_apis.rs (hunk 1 above)
Comment (joshuef) [reply src/client/client_api/map_apis.rs:895]: hmm could be. But we'd then have to refactor that API as well. Which maybe we want, but this is just tests so I'd lean on not doing that just for that

File: src/client/client_api/sequence_apis.rs (hunk 3 above)
Comment (joshuef) [reply src/client/client_api/sequence_apis.rs:791]: yeh, this is basically the same as previous, but the loop pattern matching doesnt let us (or I didn't figure out), how to get it to match against an or here.

Balance could change, but wont go to 0 during this test.

Comment (grumbach) [reply src/client/client_api/sequence_apis.rs:791]:
```suggestion
                let balance_before_delete = retry_loop_for_pattern!(client.get_balance(), Ok(bal) if *bal != Token::from_str("10")? || *bal != Token::from_str("0")?)?;
```
Wouldn't this do the trick?

Comment (grumbach) [inline src/client/client_api/sequence_apis.rs:791]:
```suggestion
        let balance_before_delete = retry_loop_for_pattern!(client.get_balance(), Ok(bal) if *bal != Token::from_str("10")? && *bal != Token::from_str("0")?)?;
```

**Pull request merged**